### PR TITLE
[PSPE] mcpeinventory

### PR DESCRIPTION
**Pull request by 7kasper**
OHh my god... This will take some time to koam into something tidy.
Anyway I will use this to keep track of some stuff that needs to be done in order to make inventory branch mergable.

Things that need to be done:
- ~~Use request channel instead of hacking into bukkit events (fake inventories and inventory update)~~
- Stuff

### CODE REVIEW

**Comment by 7kasper** [comment]
image

**Comment by MrPowerGamerBR** [comment]
wow

**Comment by NycuRO** [comment]
You fixed all inventories ? Inclusive enchantments, anvil, brewing, crafting, crafting human, chest, hopper, dropper, furnace and others?

**Comment by 7kasper** [comment]
Yess but it neess some rewrites

**Comment by NycuRO** [comment]
What is next? :D

**Comment by 7kasper** [comment]
Ok. I rewrote the fake inventories to only store position instead of the block and use plugin requests to update the fake containers / inventories.

Bescides debug and beacons (which are broken by the latest addition) I need to know what to change in order to merge this 💩 😄

**Comment by Shevchik** [comment]
There is a netcache refactor going in in main branch right now, so i will refactor mcpenew storage a bit too. After that i will check mcpeinventory branch, maybe something can be improved in it.
(Maybe some of the inventories faking should be migrated to serverside handlers instead of network?)

**Comment by 7kasper** [comment]
Yeah sure, I will try to do it like your cache when its done.
I do think the current way of inventory faking is the way to go, since you can really easily make clientside blocks and send the inventoryopen packet in the same recyclable packet array (except for beacons and double chests) but I'll see.

**Comment by 7kasper** [comment]
Updated to the new cache stuff.

**Comment by 7kasper** [comment]
In discord we decided to split this up.. again :D
So we will first do all the non- player inventory related stuff done in this branch and then move on to this PR.

File: src/protocolsupport/protocol/typeremapper/pe/inventory/PESlotRemapper.java [hunk 1]
```diff
@@ -1,13 +1,270 @@
...
+import protocolsupport.zplatform.itemstack.ItemStackWrapper;
 
 public class PESlotRemapper {
-	
+
+	public static ClientBoundPacketData remapClientBoundSlot(NetworkDataCache cache, ProtocolVersion version, ItemStackWrapper item, int slot) {
+		String locale = cache.getAttributesCache().getLocale();
+		PEInventoryCache invCache = cache.getPEInventoryCache();
+		WindowCache winCache = cache.getWindowCache();
+		switch(cache.getWindowCache().getOpenedWindow()) {
```
Comment (NycuRO) [inline src/protocolsupport/protocol/typeremapper/pe/inventory/PESlotRemapper.java]: omg big switcher. use fastutil and maps..

**Comment by 7kasper** [comment]
Okido ready to accept code reviews I think. I will update the wikis another time, they are mostly correct.
This code adds inventory support for everything and every action possible in minecraft PE and Java except two things:
* Shift clicking with item in Cursor (will fail softly).
* Lama inventories (not entirely done yet).

File: src/protocolsupport/protocol/typeremapper/pe/inventory/PESlotRemapper.java (hunk 1 above)
Comment (7kasper) [reply src/protocolsupport/protocol/typeremapper/pe/inventory/PESlotRemapper.java]: Maybe... But the remapping is all very dependend on the type of inventory opened. I could remap with windowId and slot into one integer but then we have a giant map and using that with the inventorySetItems will actually cause way worse performance.

Comment (NycuRO) [reply src/protocolsupport/protocol/typeremapper/pe/inventory/PESlotRemapper.java]: Okey.

**Comment by magicus** [comment]
GodPacket, Line 114, that's just plain invCache, right?

**Comment by 7kasper** [comment]
Correct ill change with next commit

**Comment by 7kasper** [comment]
BRB

**Comment by 7kasper** [comment]
Bump 😄

**Comment by 7kasper** [comment]
Alright chums, I'm back!

**Comment by 7kasper** [comment]
Let’s do this.

**Pull request merged**